x/protorev/keeper/posthandler.go

```diff
@@ -0,0 +1,123 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v13/x/gamm/types"
+)
+
+type SwapToBackrun struct {
+	PoolId        uint64
+	TokenOutDenom string
+	TokenInDenom  string
+}
+
+type ProtoRevDecorator struct {
+	ProtoRevKeeper Keeper
+}
+
+func NewProtoRevDecorator(protoRevDecorator Keeper) ProtoRevDecorator {
+	return ProtoRevDecorator{
+		ProtoRevKeeper: protoRevDecorator,
+	}
+}
+
+// This posthandler will first check if there were any swaps in the tx. If so, collect all of the pools, build three
+// pool routes for cyclic arbitrage, and then execute the optimal route if it exists.
+func (protoRevDec ProtoRevDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// Create a cache context to execute the posthandler such that
+	// 1. If there is an error, then the cache context is discarded
+	// 2. If there is no error, then the cache context is written to the main context
+	cacheCtx, write := ctx.CacheContext()
+
+	txGasWanted := cacheCtx.GasMeter().Limit()
+	// Ignore cases where limit is 0 (edge case for genUtil init)
+	if txGasWanted == 0 {
+		return next(ctx, tx, simulate)
+	}
+
+	// Change the gas meter to be an infinite gas meter which allows the entire posthandler to work without consuming any gas
+	cacheCtx = cacheCtx.WithGasMeter(sdk.NewInfiniteGasMeter())
+
+	// Only execute the protoRev module if it is enabled
+	if enabled, err := protoRevDec.ProtoRevKeeper.GetProtoRevEnabled(cacheCtx); err != nil || !enabled {
+		return next(ctx, tx, simulate)
+	}
+
+	// Get the max number of pools to iterate through
+	maxPoolsToIterate, err := protoRevDec.ProtoRevKeeper.GetMaxPools(cacheCtx)
+	if err != nil {
+		return next(ctx, tx, simulate)
+	}
+
+	// Find routes for every single pool that was swapped on (up to maxPoolsToIterate pools per tx)
+	swappedPools := ExtractSwappedPools(tx)
+	tradeErr := error(nil)
+	for index, swap := range swappedPools {
+		if uint64(index) >= maxPoolsToIterate {
+			break
+		}
+		// If there was is an error executing the trade, break and set tradeErr
+		if err := protoRevDec.ProtoRevKeeper.ProtoRevTrade(cacheCtx, swap); err != nil {
+			tradeErr = err
+			break
+		}
+	}
+	// If there was no error, write the cache context to the main context
+	if tradeErr == nil {
+		write()
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	} else {
+		ctx.Logger().Error("ProtoRevTrade failed with error", tradeErr)
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// ProtoRevTrade wraps around the build routes, iterate routes, and execute trade functionality to execute an cyclic arbitrage trade
+// if it exists. It returns an error if there was an error executing the trade and a boolean if the trade was executed.
+func (k Keeper) ProtoRevTrade(ctx sdk.Context, swap SwapToBackrun) error {
```

Comment:
It looks like we're not doing anything to limit the maximum number of routes iterated per block (i.e. to explicitly cap the extra amount of computation protorev can be responsible for in the worst case). 

Is this because that was hard to do or because we decided that was unnecessary for some reason?

Reply:
Correct, as of now there are only local bounds (within each tx), nothing across the block.

We should be able to add an across block bound if we want to. 

The current decision to go to local bounding was because a cheap attack vector is to make a single tx that swaps against 100+ pools that busies up protorev, so local bounding is necessary (at min, so chose this over the across block bound method).

I'd be comfortable with a lenient across block bound that allows for up to 100 swaps (or some number that goes above some target average runtime), mainly would be concerned about attacks though rather than real defi activity for the time being. @bpiv400

Reply:
@bpiv400 we've updated the code such that we now have a maximum # of txs that can be processed per block. The counter tracking this will only be incremented if the tx had at least one swap msg within it. If the number of txs has exceeded the threshold, the posthandler does nothing. Additionally, we make this a configurable parameter that can be altered via the admin account (but still has an acceptable range so that the admin account doesn't introduce an attack vector).

Reply:
update: we now constrain protorev by the number of routes traversed by block and tx.